game-core/src/main/java/games/strategy/triplea/delegate/MoveValidator.java

```diff
@@ -268,6 +268,28 @@ private static String validateCanal(final Route route, final Collection<Unit> un
     return checkLandTransports(data, player, potentialLandTransports, unitsToLandTransport).isEmpty() ? null : result;
   }
 
+  /**
+   * Simplified version of {@link #validateCanal(Route, Collection, PlayerID, GameData) validateCanal} used for route
+   * finding to check neighboring territories and avoid validating every unit. Performance of this method is critical.
+   */
+  public static boolean canAnyUnitsPassCanal(final Territory start, final Territory end, final Collection<Unit> units,
```

Comment:
This is the core of the implementation which is a simplified canal validation method that checks if the canal is controlled and at least 1 unit can pass through. This makes it much faster than a full validation.